### PR TITLE
Adding documentation for GAM

### LINKED ISSUE #1645: Add Documentation for GAM Trainers
As @sfilipi pointed out on a PR #1642, the GAM trainers don't have samples available in the dynamic API. I suggest that we add samples to show how to use these learners.

@sfilipi 's comment:
```
My new comment on every PR that touches the catalogs: how do you feel about writing a sample for this, and referencing it here :)

see: https://github.com/dotnet/machinelearning/blob/master/docs/samples/Microsoft.ML.Samples/Dynamic/SDCA.cs

and its reference:
https://github.com/dotnet/machinelearning/src/Microsoft.ML.StandardLearners/StandardLearnersCatalog.cs
```

### PR DESCRIPTION
**Pull request by rogancarr**
This PR adds documentation for using the GAM in the Microsoft.ML.Samples project.

Fixes #1645 

Awaiting PR #1691

### CODE REVIEW

File: docs/samples/Microsoft.ML.Samples/Dynamic/GAM.cs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+            var teacherRatioFeatureWeights = gamModel.GetFeatureWeights(10);
+
+            Console.WriteLine("We can see that smaller class sizes are predictive of a higher house value. Student-teacher ratios larger than ~18, lead to lower-than-average house price predictions.");
+            Console.WriteLine("Student-Teacher Ratio");
+            for (int i = 0; i < teacherRatioBinUpperBounds.Length; i++)
+            {
+                Console.WriteLine($"x < {teacherRatioBinUpperBounds[i]:0.00} => {teacherRatioFeatureWeights[i]:0.000}");
+            }
+
+            Console.WriteLine("Note that these measurements are noisy (see student-teacher ratios > xyz). Common practice is to use resampling methods to estimate confidence at each bin.");
```
Comment (rogancarr) [inline docs/samples/Microsoft.ML.Samples/Dynamic/GAM.cs]: Will update the `xyz` placeholder with the correct value once the awaited PR is in.

**Comment by rogancarr** [comment]
I'd appreciate a review of this API documentation; while this is awaiting another PR, that only affects the names and signatures of a few APIs used, and doesn't need to block a critical review.

File: docs/samples/Microsoft.ML.Samples/Dynamic/GAM.cs [hunk 2]
```diff
@@ -0,0 +1,82 @@
+﻿using Microsoft.ML.Runtime.Data;
+using Microsoft.ML.Trainers.FastTree;
+using System;
+using System.Linq;
```
Comment (rogancarr) [inline docs/samples/Microsoft.ML.Samples/Dynamic/GAM.cs]: Can drop `System.Linq`.

File: docs/samples/Microsoft.ML.Samples/Dynamic/GAM.cs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+            // Read the data
+            var data = reader.Read(dataFile);
+
+            // Step 2: Pipeline
+            // Concatenate the features to create a Feature vector.
+            // Then append a gam regressor, setting the "MedianHomeValue" column as the label of the dataset, and 
+            // the "Features" column produced by concatenation as the features column.
+            var pipeline = mlContext.Transforms.Concatenate("Features", "CrimesPerCapita", "PercentResidental", "PercentNonRetail", 
+                "CharlesRiver", "NitricOxides", "RoomsPerDwelling", "PercentPre40s", "EmploymentDistance", "HighwayDistance", "TaxRate", "TeacherRatio")
+                    .Append(mlContext.Regression.Trainers.GeneralizedAdditiveModels(labelColumn: "MedianHomeValue", featureColumn: "Features", maxBins: 16));
```
Comment (rogancarr) [inline docs/samples/Microsoft.ML.Samples/Dynamic/GAM.cs]: Somewhere, make a note that we are using very few bins to make it easy to parse the examples.

File: docs/samples/Microsoft.ML.Samples/Dynamic/GAM.cs (hunk 1 above)
Comment (rogancarr) [reply docs/samples/Microsoft.ML.Samples/Dynamic/GAM.cs]: Resolved.

File: docs/samples/Microsoft.ML.Samples/Dynamic/GAM.cs (hunk 2 above)
Comment (rogancarr) [reply docs/samples/Microsoft.ML.Samples/Dynamic/GAM.cs]: Resolved usings.

File: docs/samples/Microsoft.ML.Samples/Dynamic/GAM.cs (hunk 3 above)
Comment (rogancarr) [reply docs/samples/Microsoft.ML.Samples/Dynamic/GAM.cs]: Added a note in the comments.

File: docs/samples/Microsoft.ML.Samples/Dynamic/GAM.cs [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+            var data = reader.Read(dataFile);
+
+            // Step 2: Pipeline
+            // Concatenate the features to create a Feature vector.
+            // Then append a gam regressor, setting the "MedianHomeValue" column as the label of the dataset,
+            // the "Features" column produced by concatenation as the features column,
+            // and use a small number of bins to make it easy to visualize in the console window.
+            // For real appplications, it is recommended to start with the default number of bins.
+            var pipeline = mlContext.Transforms.Concatenate("Features", "CrimesPerCapita", "PercentResidental",
+                "PercentNonRetail", "CharlesRiver", "NitricOxides", "RoomsPerDwelling", "PercentPre40s",
```
Comment (sfilipi) [inline docs/samples/Microsoft.ML.Samples/Dynamic/GAM.cs]: indentation is off.

File: docs/samples/Microsoft.ML.Samples/Dynamic/GAM.cs [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+            var fitPipeline = pipeline.Fit(data);
+
+            // Extract the model from the pipeline
+            var gamModel = fitPipeline.LastTransformer.Model;
+
+            // Step 3: Investigate the properties of the model
+
+            // The intercept for the GAM models represent the average prediction for the training data
+            var intercept = gamModel.Intercept;
+            Console.WriteLine($"Average predicted cost: {intercept}");
```
Comment (sfilipi) [inline docs/samples/Microsoft.ML.Samples/Dynamic/GAM.cs]: might want to put it in commented the output of this WriteLine.

File: docs/samples/Microsoft.ML.Samples/Dynamic/GAM.cs (hunk 4 above)
Comment (rogancarr) [reply docs/samples/Microsoft.ML.Samples/Dynamic/GAM.cs]: Resolved.

File: docs/samples/Microsoft.ML.Samples/Dynamic/GAM.cs (hunk 5 above)
Comment (rogancarr) [reply docs/samples/Microsoft.ML.Samples/Dynamic/GAM.cs]: Resolved.

File: docs/samples/Microsoft.ML.Samples/Dynamic/GAM.cs [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+                        labelColumn: "MedianHomeValue", featureColumn: "Features", maxBins: 16));
+            var fitPipeline = pipeline.Fit(data);
+
+            // Extract the model from the pipeline
+            var gamModel = fitPipeline.LastTransformer.Model;
+
+            // Step 3: Investigate the properties of the model
+
+            // The intercept for the GAM models represent the average prediction for the training data
+            var intercept = gamModel.Intercept;
```
[the remark this answers is not in the document]
Comment (rogancarr) [reply docs/samples/Microsoft.ML.Samples/Dynamic/GAM.cs]: The `TreeTrainerCatalog` definition of `GeneralizedAdditiveModels` returns a `RegressionGamTrainer`, which returns a `RegressionPredictionTransformer<Trainers.FastTree.RegressionGamPredictor>`. No `using` is necessary.

File: docs/samples/Microsoft.ML.Samples/Dynamic/GAM.cs [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+            Console.WriteLine($"Average predicted cost: {intercept}");
+
+            // Each feature represents the deviation from the average prediction as a function of the feature value
+            
+            // Let's take a look at the TeacherRatio
```
[the remark this answers is not in the document]
Comment (rogancarr) [reply docs/samples/Microsoft.ML.Samples/Dynamic/GAM.cs]: Done. I fetch the features array from the `IDataView` used to train the model.

File: docs/samples/Microsoft.ML.Samples/Dynamic/GAM.cs [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+            var teacherRatioBinUpperBounds = gamModel.GetFeatureBinUpperBounds(10);
```
[the remark this answers is not in the document]
Comment (rogancarr) [reply docs/samples/Microsoft.ML.Samples/Dynamic/GAM.cs]: I think we should. And I think that the `GAMPredictor` should also maintain the column names internally. Would you be open to fixing this in a different PR?

File: docs/samples/Microsoft.ML.Samples/Dynamic/GeneralizedAdditiveModels.cs [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+            // model responds to the variable on a global level. For the student-teacher-ratio variable, we can see
+            // that smaller class sizes are predictive of a higher house value, while student-teacher ratios higher 
+            // than about 18 lead to lower predictions in house value. This makes intuitive sense, as smaller class 
+            // sizes are desirable and also indicative of better-funded schools, which could make buyers likely to
+            // pay more for the house.
+            //
+            // Another thing to notice is that these feature functions can be noisy. See student-teacher ratios > 21.05.
+            // Common practice is to use resampling methods to estimate a confidence interval at each bin. This will
+            // help to determine if the effect is real or just sampling noise. See for example 
+            // Tan, Caruana, Hooker, and Lou. "Distill-and-Compare: Auditing Black-Box Models Using Transparent Model
```
Comment (sfilipi) [inline docs/samples/Microsoft.ML.Samples/Dynamic/GeneralizedAdditiveModels.cs]: link

Comment (rogancarr) [reply docs/samples/Microsoft.ML.Samples/Dynamic/GeneralizedAdditiveModels.cs]: Resolved.

File: docs/samples/Microsoft.ML.Samples/Dynamic/GeneralizedAdditiveModels.cs [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+            // than about 18 lead to lower predictions in house value. This makes intuitive sense, as smaller class 
+            // sizes are desirable and also indicative of better-funded schools, which could make buyers likely to
+            // pay more for the house.
+            //
+            // Another thing to notice is that these feature functions can be noisy. See student-teacher ratios > 21.05.
+            // Common practice is to use resampling methods to estimate a confidence interval at each bin. This will
+            // help to determine if the effect is real or just sampling noise. See for example 
+            // Tan, Caruana, Hooker, and Lou. "Distill-and-Compare: Auditing Black-Box Models Using Transparent Model 
+            // Distillation." <a href='https://arxiv.org/abs/1710.06169'>arXiv:1710.06169</a>."
+            Console.WriteLine(string.Empty);
```
Comment (Ivanidzo4ka) [inline docs/samples/Microsoft.ML.Samples/Dynamic/GeneralizedAdditiveModels.cs]: > Console.WriteLine(string.Empty); [](start = 11, length = 33)

This is technically correct, but weird way to move to new line.
You can just use Console.Writeline();
All across this file.

**Pull request merged**